// @flagship.io/js-sdk/dist/visitor/NoConsentStrategy.js
import { FLAG_VISITOR_EXPOSED, METHOD_DEACTIVATED_CONSENT_ERROR } from '../enum/index';
import { logInfo, sprintf } from '../utils/utils';
import { DefaultStrategy } from './DefaultStrategy';
export class NoConsentStrategy extends DefaultStrategy {
    async lookupHits() {
        //
    }
    async lookupVisitor() {
        //
    }
    async cacheVisitor() {
        //
    }
    async cacheHit() {
        //
    }
    fetchVisitorCampaigns() {
        return [];
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async sendHit(_hit) {
        this.log('sendHit');
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async sendHits(_hits) {
        this.log('sendHits');
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async sendActivate(_flagDto, _defaultValue) {
        //
    }
    async visitorExposed() {
        this.log(FLAG_VISITOR_EXPOSED);
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async sendTroubleshootingHit(_hit) {
        //
    }
    log(methodName) {
        logInfo(this.config, sprintf(METHOD_DEACTIVATED_CONSENT_ERROR, methodName, this.visitor.visitorId), methodName);
    }
}
