// @flagship.io/js-sdk/dist/enum/FSFetchReasons.js
/**
 * Enum representing the reasons for fetching Flags.
 */
export var FSFetchReasons;
(function (FSFetchReasons) {
    /**
     * Indicates that a context has been updated or changed.
     */
    FSFetchReasons["UPDATE_CONTEXT"] = "UPDATE_CONTEXT";
    /**
     * Indicates that the XPC method 'authenticate' has been called.
     */
    FSFetchReasons["AUTHENTICATE"] = "AUTHENTICATE";
    /**
     * Indicates that the XPC method 'unauthenticate' has been called.
     */
    FSFetchReasons["UNAUTHENTICATE"] = "UNAUTHENTICATE";
    /**
     * Indicates that fetching flags has failed.
     */
    FSFetchReasons["FETCH_ERROR"] = "FETCH_ERROR";
    /**
     * Indicates that flags have been fetched from the cache.
     */
    FSFetchReasons["READ_FROM_CACHE"] = "READ_FROM_CACHE";
    /**
     * Indicates that the visitor has been created.
     */
    FSFetchReasons["VISITOR_CREATED"] = "VISITOR_CREATED";
    /**
     * Indicates that there is no specific reason for fetching flags.
     */
    FSFetchReasons["NONE"] = "NONE";
})(FSFetchReasons || (FSFetchReasons = {}));
